src/router/open-plc-utils/ether/efeu.1
.TH efeu 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
efeu - HomePlug AV Packet Reflector

.SH SYNOPSIS
.BR efeu 
.RI [ options ] 

.SH DESCRIPTION
Read and reflect 802.3 Ethernet messages addressed to this host or the broadcast address.
Do not display reflected frames.

.PP
The program is one example of a simple packet listener.
Engineers and software developers can extend it to build custom sniffers or servers.
The framework is present but additional features must be added.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
-\fBe \fIethertype\fR
The protocol (or ethertype) of interest expressed as a 16-bit hexadecimal integer.
Prefix "\fB0x\fR" is optional.
This filters out frames having any other ethertype.
The default is \fBETH_P_802_2\fR as defined in \fBif_ether.h\fR.

.TP
-\fBi \fIinterface\fR
Names the host ethernet interface to watch.
The named interface, such as eth0, eth1 and so on, must be connected to a HomePlug AV ethernet-to-powerline bridge adapter to receive any HomePlug AV traffic.
The default interface is \fBeth1\fR.

.TP
.RB - q
Quiet mode.
Suppress the printing of the normal message descriptions on \fBstdout\fR.

.TP
-\fBt \fImilliseconds\fR
Read timeout in milliseconds.
Values range from 0 through UINT_MAX.
This is the maximum time allowed for a response.
The default is 50 milliseconds.

.TP
.RB - v
Verbose mode.
Displays incoming frames in hexadecimal dump format on \fBstdout\fR.
This option is independent of option \fB-a\fR which prints VS_ARPC message text on stderr.

.TP
-\fB?\fR, --\fBhelp\fR
Print program help summary on stdout.
This option takes precedence over other options on the command line.
.Tp
-\fB!\fR, --\fBversion\fR
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.
See the HomePlug AV Specification for a description of all HomePlug AV message types.
See the Qualcomm Atheros INT6000 Firmware Reference Manual for a description of Atheros Vendor Specific message types.

.SH EXAMPLES
The following command monitors the default interface \fBeth1\fR for HomePlug AV traffic and dumps detected frames on the console as they occur.
Terminate the foreground process by typing \fB^C\fR or \fB^Z\fR at the console.
This program is usually run in a separate console window when performing other toolkit operations.

.PP
   # efeu

.PP
The following command monitors host interface \fBeth4\fR for incoming HomePlug AV frames and dumps writes them to file \fBefeu.log\fR.
Frames are printed in hexadecimal dump format because option \fB-v\fR is present.
Terminate the foreground process by typing ^C or ^Z at the console.

.PP
   # efeu -v -i eth4 > efeu.log

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information is not available.
Qualcomm Atheros reserves the right to modify management message structure and content in future firmware releases without any obligation to notify or compensate users of this program.

.SH SEE ALSO
.BR plc ( 1 ),
.BR efru ( 1 ),
.BR efsu ( 1 ),
.BR hpav ( 1 ),
.BR mme ( 1 )

.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>



src/router/open-plc-utils/ether/efeu.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or 
 *   without modification, are permitted (subject to the limitations 
 *   in the disclaimer below) provided that the following conditions 
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright 
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above 
 *     copyright notice, this list of conditions and the following 
 *     disclaimer in the documentation and/or other materials 
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of 
 *     its contributors may be used to endorse or promote products 
 *     derived from this software without specific prior written 
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE 
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE 
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER 
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) 
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   edeu.c - Ethernet Frame Echo Utility
 *
 *   Listen for incoming Ethernet II frames and write frame data to
 *   stdout as a continuous binary data stream. Since frame data can
 *   contain anything, users should send program output to a file or
 *   pipe it through a filter program to prevent corruption of the
 *   terminal session. If all incoming data is ASCII then directing
 *   stdout to the console should not cause any problems.
 *
 *   this program can be used as a data target when testing AR6405
 *   UART applications; use program edsu to send files from a remote
 *   host and this program to display or save them;
 *
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/


/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <fcntl.h>
#include <stdio.h>
#include <limits.h>
#include <ctype.h>
#include <unistd.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#ifdef __linux__
#include <net/if.h>
#include <net/if_arp.h>
#include <netpacket/packet.h>
#include <signal.h>
#endif

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/error.h"
#include "../ether/channel.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/hexdump.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/uintspec.c"
#include "../tools/memswap.c"
#include "../tools/todigit.c"
#include "../tools/error.c"
#endif

#ifndef MAKEFILE
#include "../ether/channel.c"
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/sendpacket.c"
#include "../ether/readpacket.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define EFEU_INTERFACE "PLC"
#define EFEU_ETHERTYPE ETH_P_802_2

/*====================================================================*
 *
 *   int main (int argc, char * argv[]);
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	struct ethernet_frame frame;
	signed length;
	static char const * optv [] =
	{
		"e:i:qt:v",
		PUTOPTV_S_DIVINE,
		"Qualcomm Atheros Ethernet Frame Echo Utility",
		"e x\tethertype is (x) [" LITERAL (EFEU_ETHERTYPE) "]",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"q\tsuppress normal output",
		"t n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_FOREVER) "]",
		"v\tverbose messages on stdout",
		(char const *) (0)
	};
	signed c;
	channel.type = EFEU_ETHERTYPE;
	channel.timeout = CHANNEL_FOREVER;
	if (getenv (EFEU_INTERFACE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (EFEU_INTERFACE));

#else

		channel.ifname = strdup (getenv (EFEU_INTERFACE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'e':
			channel.type = (uint16_t)(basespec (optarg, 16, sizeof (channel.type)));
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			break;
		case 't':
			channel.timeout = (signed)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc)
	{
		error (1, ECANCELED, ERROR_TOOMANY);
	}
	openchannel (&channel);
	while ((length = readpacket (&channel, &frame, sizeof (frame))) > 0)
	{
		memswap (&frame.frame_dhost, &frame.frame_shost, sizeof (frame.frame_dhost));
		if (sendpacket (&channel, &frame, length) != length)
		{
			error (1, errno, CHANNEL_CANTSEND);
		}
	}
	closechannel (&channel);
	return (0);
}

